[SPARK-19799][SQL] Support WITH clause in subqueries

peter-toth opened the pull request:
## What changes were proposed in this pull request?

This PR  adds support of `WITH` clause within a subquery so this query becomes valid:
  ```
  SELECT max(c) FROM (
    WITH t AS (SELECT 1 AS c)
    SELECT * FROM t
  )
 ```

## How was this patch tested?

Added new UTs.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

dongjoon-hyun commented:
ok to test

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/11606/
Test PASSed.

SparkQA commented:
**[Test build #106360 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106360/testReport)** for PR 24831 at commit [`741c727`](https://github.com/apache/spark/commit/741c7271fe4b6a5ca4e66125a2a84e8755616993).

SparkQA commented:
**[Test build #106360 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106360/testReport)** for PR 24831 at commit [`741c727`](https://github.com/apache/spark/commit/741c7271fe4b6a5ca4e66125a2a84e8755616993).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106360/
Test PASSed.

dongjoon-hyun commented:
cc @gatorsmile since this is a part of PostgreSQL feature parity.

SparkQA commented:
**[Test build #106461 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106461/testReport)** for PR 24831 at commit [`0b516fc`](https://github.com/apache/spark/commit/0b516fccc95e3b6c8d883afb3e160847f4ae7d4c).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/11704/
Test PASSed.

SparkQA commented:
**[Test build #106461 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106461/testReport)** for PR 24831 at commit [`0b516fc`](https://github.com/apache/spark/commit/0b516fccc95e3b6c8d883afb3e160847f4ae7d4c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106461/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/11782/
Test PASSed.

SparkQA commented:
**[Test build #106541 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106541/testReport)** for PR 24831 at commit [`ca27852`](https://github.com/apache/spark/commit/ca27852b884eb7c0061627493919a33fc3fe41f6).

SparkQA commented:
**[Test build #106541 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106541/testReport)** for PR 24831 at commit [`ca27852`](https://github.com/apache/spark/commit/ca27852b884eb7c0061627493919a33fc3fe41f6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106541/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12012/
Test PASSed.

SparkQA commented:
**[Test build #106793 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106793/testReport)** for PR 24831 at commit [`d76a265`](https://github.com/apache/spark/commit/d76a265545a4273f69fd9eb3d46afa7bfafb763a).

SparkQA commented:
**[Test build #106793 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106793/testReport)** for PR 24831 at commit [`d76a265`](https://github.com/apache/spark/commit/d76a265545a4273f69fd9eb3d46afa7bfafb763a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106793/
Test PASSed.

maropu commented on sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:
```diff
@@ -412,7 +412,7 @@ queryPrimary
     | TABLE multipartIdentifier                                             #table
     | inlineTable                                                           #inlineTableDefault1
-    | '(' queryNoWith  ')'                                                  #subquery
+    | '(' query  ')'                                                        #subquery
```
nit: remove an unnecessary single space.

peter-toth replied:
ok, done

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -212,31 +212,96 @@ class Analyzer(
-   * Analyze cte definitions and substitute child plan with analyzed cte definitions.
+   * Analyze WITH nodes and substitute child plan with CTE definitions.
    */
   object CTESubstitution extends Rule[LogicalPlan] {
```
How about making a new file for this rule? cc: @gatorsmile

peter-toth replied:
ok moved

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -212,31 +212,96 @@ class Analyzer(
-        }
-      case other => other
-    }
+    def apply(plan: LogicalPlan): LogicalPlan = traverse(plan, false)
```
I feel the current logic is a little hard to understand because of the legacy mode.
I think the existing logic for WITH is pretty simple, so how about keeping the existing one and adding a new logic for subqueries like this?
```
  object CTESubstitution extends Rule[LogicalPlan] {
    def apply(plan: LogicalPlan): LogicalPlan = if (!SQLConf.get.legacyCTESubstitutionEnabled) {
      traverseAndsubstituteCTE(plan)
    } else {
      legacyTraverseAndsubstituteCTE(plan)
    }
```

peter-toth replied:
Thanks @maropu for the review! I changed as you suggested.

maropu commented on sql/core/src/test/resources/sql-tests/inputs/cte.sql:
```diff
@@ -28,6 +28,177 @@ FROM   CTE1 t1
+);
+
+-- CTE in CTE definition shadows outer (legacy substitution)
+SET spark.sql.legacy.cte.substitution.enabled=true;
```
Could we split these tests into two separate files: the legacy one and non-legacy one?

peter-toth replied:
ok, done

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12259/
Test PASSed.

SparkQA commented:
**[Test build #107069 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107069/testReport)** for PR 24831 at commit [`9096c4d`](https://github.com/apache/spark/commit/9096c4dc0098f3a2bd034e3eaec33f23541cabef).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12261/
Test PASSed.

SparkQA commented:
**[Test build #107073 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107073/testReport)** for PR 24831 at commit [`fd9ca42`](https://github.com/apache/spark/commit/fd9ca4226e72f8086353cae779cb51b3e4216586).

mgaido91 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala:
```diff
@@ -0,0 +1,122 @@
+   *   )
+   * @param plan  the plan to be traversed
+   * @param inTraverse whether the current traverse is called from another traverse, only in this
+   *                   can name collision occur
```
```suggestion
   *                   case name collision can occur
```

mgaido91 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala:
```diff
@@ -0,0 +1,122 @@
+
+      case o if inTraverse =>
+        o.transformExpressions {
+          case e: SubqueryExpression => e.withNewPlan(traverseAndSubstituteCTE(e.plan, inTraverse))
```
`inTraverse` can be replaced with `true`, right?

peter-toth commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala:
```diff
@@ -0,0 +1,122 @@
+      plan: LogicalPlan,
+      cteName: String,
+      ctePlan: => LogicalPlan): LogicalPlan = {
+    lazy val lazyCTEPlan = ctePlan
```
Thanks @mgaido91! It is not needed indeed.

peter-toth replied:
I remember now why I used lazy here. A CTE definition can be used multiple times in `WITH` but the call by name parameter (`ctePlan  = traverseAndSubstituteCTE(...)`) should be executed only once.
But now I believe it is better to use lazy outside of `substituteCTE` than inside, please review my commit https://github.com/apache/spark/pull/24831/commits/7d6910599134dc992bc841dfd3eff6f46d36647b.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12264/
Test PASSed.

SparkQA commented:
**[Test build #107079 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107079/testReport)** for PR 24831 at commit [`39b9596`](https://github.com/apache/spark/commit/39b959604ef026550a50d8cef96f66801dfaa03f).

SparkQA commented:
**[Test build #107069 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107069/testReport)** for PR 24831 at commit [`9096c4d`](https://github.com/apache/spark/commit/9096c4dc0098f3a2bd034e3eaec33f23541cabef).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107069/
Test PASSed.

SparkQA commented:
**[Test build #107073 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107073/testReport)** for PR 24831 at commit [`fd9ca42`](https://github.com/apache/spark/commit/fd9ca4226e72f8086353cae779cb51b3e4216586).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107073/
Test PASSed.

SparkQA commented:
**[Test build #107079 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107079/testReport)** for PR 24831 at commit [`39b9596`](https://github.com/apache/spark/commit/39b959604ef026550a50d8cef96f66801dfaa03f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107079/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12267/
Test PASSed.

SparkQA commented:
**[Test build #107084 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107084/testReport)** for PR 24831 at commit [`7d69105`](https://github.com/apache/spark/commit/7d6910599134dc992bc841dfd3eff6f46d36647b).

SparkQA commented:
**[Test build #107084 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107084/testReport)** for PR 24831 at commit [`7d69105`](https://github.com/apache/spark/commit/7d6910599134dc992bc841dfd3eff6f46d36647b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107084/
Test PASSed.

maropu commented on sql/core/src/test/resources/sql-tests/inputs/cte-legacy.sql:
```diff
@@ -0,0 +1,59 @@
+  )
+);
+
+SET spark.sql.legacy.cte.substitution.enabled=false;
```
you don't need this.

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala:
```diff
@@ -0,0 +1,122 @@
+    traverseAndSubstituteCTE(plan, false)
+  }
+
+  def legacyTraverseAndsubstituteCTE(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUp {
```
private?

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala:
```diff
@@ -0,0 +1,122 @@
+            substituteCTE(currentPlan, cteName, substitutedCTEPlan)
+        }
+
+      case o if inTraverse =>
```
Why we need this flag? Some tests would fail if we have not?

peter-toth replied:
No it would not, but I wanted to do CTE substitution in the current plan only (not in the subqueries) if it is safe. (CTE substitution will run for subqueries later anyway.)

maropu commented:
@peter-toth Can we split this pr into two parts cleanly: 1. the sub-query support and 2. the behaviour change? Also, I think we need to update the migration guide for the part 2.

peter-toth committed:
[SPARK-19799][SQL] Support WITH clause in subqueries

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12281/
Test PASSed.

SparkQA commented:
**[Test build #107101 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107101/testReport)** for PR 24831 at commit [`cff4bec`](https://github.com/apache/spark/commit/cff4becc1179d2c980e0b5b93ea26c31100824ee).

peter-toth commented:
> @peter-toth Can we split this pr into two parts cleanly: 1. the sub-query support and 2. the behaviour change? Also, I think we need to update the migration guide for the part 2. cc: @gatorsmile

All right, I dropped the changes that relate to order of substitution and will do it in another ticket. What remained here is just the `WITH` support in subqueries.

maropu commented:
Yea, thanks!

peter-toth commented:
> Yea, thanks!

Very welcome.
I moved substitution order related changes to  https://github.com/apache/spark/pull/25029, but it is still WIP and this one should be reviewed first.

SparkQA commented:
**[Test build #107101 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107101/testReport)** for PR 24831 at commit [`cff4bec`](https://github.com/apache/spark/commit/cff4becc1179d2c980e0b5b93ea26c31100824ee).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107101/
Test FAILed.

peter-toth committed:
minor fix

SparkQA commented:
**[Test build #107106 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107106/testReport)** for PR 24831 at commit [`85e39d4`](https://github.com/apache/spark/commit/85e39d498e8f1fd08a22bb856b5e736ce652c24a).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12287/
Test PASSed.

SparkQA commented:
**[Test build #107106 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107106/testReport)** for PR 24831 at commit [`85e39d4`](https://github.com/apache/spark/commit/85e39d498e8f1fd08a22bb856b5e736ce652c24a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107106/
Test PASSed.

mgaido91 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala:
```diff
@@ -0,0 +1,52 @@
+/**
+ * Analyze WITH nodes and substitute child plan with CTE definitions.
+ */
+object CTESubstitution extends Rule[LogicalPlan] {
```
can we avoid moving the class, in order to keep the diff smaller?

peter-toth replied:
The idea of moving the rule to a separate file came from here: https://github.com/apache/spark/pull/24831#discussion_r298864911
, but I think you are right @mgaido because since that we cut the scope and split the PR. Maybe the other part (https://github.com/apache/spark/pull/25029) could extract the rule to a separate file as that one makes the rule a bit more complicated. Does that work for you @maropu?

mgaido91 replied:
Yes, that was what I meant, we can move the rule in the other PR which refactors it more thoroughly.

mgaido91 commented on sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:
```diff
@@ -412,7 +412,7 @@ queryPrimary
     | fromStatement                                                         #fromStmt
     | TABLE multipartIdentifier                                             #table
     | inlineTable                                                           #inlineTableDefault1
-    | '(' queryNoWith  ')'                                                  #subquery
```
can we also remove `queryNoWith` since it has only 1 usage?

peter-toth replied:
Thanks good idea!

mgaido91 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -211,37 +211,6 @@ class Analyzer(
-      plan resolveOperatorsUp {
-        case UnresolvedRelation(Seq(table)) if resolver(cteName, table) =>
-          ctePlan
-        case u: UnresolvedRelation =>
```
why did you remove this?

peter-toth replied:
I don't think this like does anything nor `UnresolvedRelation` can have an expression so I thought it is safe and good idea to remove it. Please correct me if I'm wrong.

mgaido91 replied:
yes, I think you're right, I was just curious about the reason of this change

peter-toth committed:
fix review findings

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12338/
Test PASSed.

SparkQA commented:
**[Test build #107168 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107168/testReport)** for PR 24831 at commit [`837c776`](https://github.com/apache/spark/commit/837c7763999b1aff30c87b2a25e3dc8ed633abf8).

mgaido91 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -226,20 +226,20 @@ class Analyzer(
+    private def substituteCTE(
+        plan: LogicalPlan,
+        cteName: String,
+        ctePlan: LogicalPlan): LogicalPlan =
```
may you please keep the  `{`. I think in general we are trying to always use them in order to help readability.

mgaido91 reviewed (commented):
only a style comment, otherwise LGTM, thanks!

mgaido91 commented on sql/core/src/test/resources/sql-tests/results/cte.sql.out:
```diff
@@ -98,16 +98,202 @@ struct<x:int>
+-- !query 20 schema
+struct<>
+-- !query 20 output
+org.apache.spark.sql.AnalysisException
```
Just a question, is this going to be addressed in the PR which allows recursive subqueries or is this an invalid query?

peter-toth replied:
I have a WIP PR open https://github.com/apache/spark/pull/23531 that would add support for recursive queries (and subqueries and subquery expressions too), but using recursive reference in a subquery is not allowed according to the SQL standard. 

But, this PR should be accepted first then could come https://github.com/apache/spark/pull/25029 and https://github.com/apache/spark/pull/23531

Actually I think I'm removing the test `WITH r AS (SELECT * FROM r) SELECT * FROM r;` because there is already a similar one in `cte.sql` and moving the `WITH r AS (SELECT (SELECT * FROM r)) SELECT * FROM r;` next to the existing one.

peter-toth committed:
fix review findings 2

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12345/
Test PASSed.

SparkQA commented:
**[Test build #107175 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107175/testReport)** for PR 24831 at commit [`d0c57c8`](https://github.com/apache/spark/commit/d0c57c8994da0a1cb9c6bd479ff26bafad325b14).

peter-toth committed:
remove duplicate test

SparkQA commented:
**[Test build #107178 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107178/testReport)** for PR 24831 at commit [`9ec6eaf`](https://github.com/apache/spark/commit/9ec6eaf3f1b992ae3c0101464483f3a7c9c8f70a).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12348/
Test PASSed.

SparkQA commented:
**[Test build #107168 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107168/testReport)** for PR 24831 at commit [`837c776`](https://github.com/apache/spark/commit/837c7763999b1aff30c87b2a25e3dc8ed633abf8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107168/
Test PASSed.

SparkQA commented:
**[Test build #107175 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107175/testReport)** for PR 24831 at commit [`d0c57c8`](https://github.com/apache/spark/commit/d0c57c8994da0a1cb9c6bd479ff26bafad325b14).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107175/
Test PASSed.

SparkQA commented:
**[Test build #107178 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107178/testReport)** for PR 24831 at commit [`9ec6eaf`](https://github.com/apache/spark/commit/9ec6eaf3f1b992ae3c0101464483f3a7c9c8f70a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107178/
Test PASSed.

maropu commented on sql/core/src/test/resources/sql-tests/results/cte.sql.out:
```diff
@@ -88,26 +98,192 @@ struct<c1:int,c2:int>
+struct<1:int>
+-- !query 12 output
+1
+
```
This result is different from the pg one;
```

postgres=# WITH
postgres-#   t AS (SELECT 1),
postgres-#   t2 AS (
postgres(#     WITH t AS (SELECT 2)
postgres(#     SELECT * FROM t
postgres(#   )
postgres-# SELECT * FROM t2;
 ?column? 
----------
        2
(1 row)
```
This will be address in the following  #25029?

dongjoon-hyun replied:
I also agree that this is inevitable in this PR. (cc @gatorsmile ).

peter-toth replied:
Yes, after https://github.com/apache/spark/pull/25029 it will return `2` (https://github.com/apache/spark/pull/25029/files#diff-fc515a5db268d29b08b80f5eb8202026R145)

dongjoon-hyun reviewed (approved):
+1, LGTM. Thank you, @peter-toth , @maropu , @mgaido91 .
The original PR is split into two (this and #25029) according to the review comment.
This is a new feature at Spark 3.0.0 and will be consistent with PostgreSQL soon.
Merged to master to move forward.

peter-toth commented:
Thanks @dongjoon-hyun, @maropu, @mgaido91 for the review! I will prepare https://github.com/apache/spark/pull/25029 for review soon.
